File: bin/convert_grades.dart
import 'dart:io';

void main() async {
  final inputFile = File('year_1404_grades.csv');
  final outputFile = File('correct_grades_1404.csv');

  if (!await inputFile.exists()) {
    print('Error: input file "year_1404_grades.csv" not found.');
    return;
  }

  final lines = await inputFile.readAsLines();
  final newLines = <String>[];

  // A map to hold the three values for each day/shift combination
  final Map<String, List<String>> dayShiftValues = {};

  for (final line in lines) {
    if (line.trim().isEmpty) continue;
    final parts = line.split(',');
    if (parts.length == 5) {
      final year = parts[0].trim();
      final month = parts[1].trim();
      final day = parts[2].trim();
      final shift = parts[3].trim();
      final value = parts[4].trim();
      final key = '$year,$month,$day,$shift';

      dayShiftValues.putIfAbsent(key, () => []).add(value);
    }
  }

  // Process the collected values
  dayShiftValues.forEach((key, values) {
    final parts = key.split(',');
    final year = parts[0];
    final month = parts[1];
    final day = parts[2];
    final shift = parts[3];

    if (values.length == 3) {
      // Assuming the order is Tailing, Product, Feed based on previous logic
      newLines.add('$year,$month,$day,$shift,باطله,${values[0]}');
      newLines.add('$year,$month,$day,$shift,محصول,${values[1]}');
      newLines.add('$year,$month,$day,$shift,خوراک,${values[2]}');
    } else if (values.length == 2) {
      // Handle cases with only 2 values if necessary, maybe assign to product/feed
      newLines.add('$year,$month,$day,$shift,محصول,${values[0]}');
      newLines.add('$year,$month,$day,$shift,خوراک,${values[1]}');
    } else if (values.length == 1) {
      // Handle cases with only 1 value
      newLines.add('$year,$month,$day,$shift,خوراک,${values[0]}');
    }
  });

  await outputFile.writeAsString(newLines.join('\n'));
  print('Conversion complete! New file created: "correct_grades_1404.csv"');
}

File: scripts/convert_grades.py
import csv

def convert_grades(input_file, output_file):
    """
    Reads the grade data, converts each line into three separate lines for each grade type,
    and writes the result to a new CSV file.
    """
    try:
        with open(input_file, 'r', encoding='utf-8') as infile, \
             open(output_file, 'w', newline='', encoding='utf-8') as outfile:
            
            reader = csv.reader(infile)
            writer = csv.writer(outfile)

            processed_shifts = set()

            for row in reader:
                if len(row) != 5:
                    continue  # Skip incorrect rows

                year, month, day, shift, value = row
                
                # Create a unique key for the day and the original shift value
                day_key = f"{year}-{month}-{day}"
                
                # Since the original file seems to have one value per shift (1, 2, 3) representing
                # different grade types, we'll process each shift number for a given day only once.
                # We'll create a unique key for the day and shift number to avoid duplicates.
                
                shift_key = f"{day_key}-{shift}"
                if shift_key in processed_shifts:
                    continue
                
                # Based on the previous logic, shift 1=Tailing, 2=Product, 3=Feed
                # Let's apply this logic to create three records for each shift
                
                try:
                    # We assume the file contains rows for shift 1, 2, and 3 for each day
                    # This script will create the correct mapping
                    
                    # For simplicity, we'll create three records for each entry.
                    # This logic assumes the value is the same for all three, which is not ideal,
                    # but it corrects the format issue.
                    # A better approach would be to have the correct data.
                    # Let's try a direct mapping based on the shift number in the row.
                    
                    grade_type = ''
                    if shift == '1':
                        grade_type = 'باطله'
                    elif shift == '2':
                        grade_type = 'محصول'
                    elif shift == '3':
                        grade_type = 'خوراک'
                    else:
                        continue # Skip if shift is not 1, 2, or 3
                        
                    writer.writerow([year, month, day, shift, grade_type, value])
                    
                except ValueError:
                    continue # Skip rows with non-integer values
                    
            print(f"Conversion successful. Corrected data saved to '{output_file}'")

    except FileNotFoundError:
        print(f"Error: The file '{input_file}' was not found.")

if __name__ == "__main__":
    convert_grades('year_1404_grades.csv', 'final_correct_grades.csv') 